Write a primer on UI testing in vue using vue-testing-library

Firebrandt opened the pull request:
I wrote a primer on UI testing in Vue using vue-testing library. 

This is the PR I'll use to get feedback on various iterations, resolve conflicts, and eventually merge my page into the wiki.

Firebrandt commented:
Currently I just have a rough outline so that I can get some feedback from my A2 partner/reviewer Nealon about:
- Whether to include some content I'm unsure about (want to keep unnecessary details out of this short primer, but some might be worth including).
- The content structure -- does it flow logically? Should sections be trimmed?
- And get answers to the Qs I included in the outline.

Fight-Fighter commented:
Hey Michael it's Nealon, here are my thoughts on your rough commit so far:
1. **Should I clarify any assumptions I make about the reader, or can I just assume they have npm and stuff?**
Never assume things about the readers. Not everybody knows the proper amount about npm and package installation, speaking from experience. If the reader needs to download something specific or structure file content and folders a certain way, be absolutely clear and allow for no ambiguity.
2. **Content Structure**:
Pretty good start on the structure! It makes sense to go straight from the introduction to the What and then Why. However doesn't it make more sense to display the "Things to Watch out For" part *after* you tell the user how to implement UI tests? Also, I know this is just a rough outline but perhaps you could list more possible use cases of UI tests to fleshen out the structure a bit? I'm worried it might be too short. Lastly, be sure that if Jest is required for this, you give the reader's an easy tool to figure out how to integrate your instructions with Jest's recognition of the unit tests. I know you posted a link, but just be sure.
3. **Content to include that you're unsure about**
-Definitely include the common pitfalls section, as stated before.
-Handy to explain why the library only allows certain queries, again you don't want any ambiguity for the reader, they should know the limits of what they can accomplish with the libraries you are providing them.
-I'd say, after you link documentation for additional query methods, you point out some that might be commonly useful, or clear up some with confusing descriptions. I know I've been confused by documentation before. It helps to write this so that a beginner programmer can understand.

**LASTLY**
Remember to edit a link to your .md file in a higher-level directory file!

Firebrandt commented:
> Hey Michael it's Nealon, here are my thoughts on your rough commit so far:
> 
>     1. **Should I clarify any assumptions I make about the reader, or can I just assume they have npm and stuff?**
>        Never assume things about the readers. Not everybody knows the proper amount about npm and package installation, speaking from experience. If the reader needs to download something specific or structure file content and folders a certain way, be absolutely clear and allow for no ambiguity.
> 
>     2. **Content Structure**:
>        Pretty good start on the structure! It makes sense to go straight from the introduction to the What and then Why. However doesn't it make more sense to display the "Things to Watch out For" part _after_ you tell the user how to implement UI tests? Also, I know this is just a rough outline but perhaps you could list more possible use cases of UI tests to fleshen out the structure a bit? I'm worried it might be too short. Lastly, be sure that if Jest is required for this, you give the reader's an easy tool to figure out how to integrate your instructions with Jest's recognition of the unit tests. I know you posted a link, but just be sure.
> 
>     3. **Content to include that you're unsure about**
>        -Definitely include the common pitfalls section, as stated before.
>        -Handy to explain why the library only allows certain queries, again you don't want any ambiguity for the reader, they should know the limits of what they can accomplish with the libraries you are providing them.
>        -I'd say, after you link documentation for additional query methods, you point out some that might be commonly useful, or clear up some with confusing descriptions. I know I've been confused by documentation before. It helps to write this so that a beginner programmer can understand.
> 
> 
> **LASTLY** Remember to edit a link to your .md file in a higher-level directory file!

Thanks for giving me some feedback before I got in deep and wrote the wrong thing! I'll start working on it.

Firebrandt commented:
>  ... However doesn't it make more sense to display the "Things to Watch out For" part after you tell the user how to implement UI tests?

Yeah, it does. I'll fix that in the first draft.

> Also, I know this is just a rough outline but perhaps you could list more possible use cases of UI tests to fleshen out the structure a bit?

I'll clarify and explain the main usage some more in the first draft -- and I'll make sure to include two examples of UI interactions that are worth testing, to get the point across.

> I'm worried it might be too short. 
Right -- although I'm intending for this to be a very quick primer on UI testing. For detailled explanations, it's best that readers go to the official docs (which are definitely gonna be better at explaining things long-form), and I do point that out I believe.

Also the length is comparable to other entries currently on the SEWiki. But a little extra content doesn't hurt, so long as it's valuable.


> Lastly, be sure that if Jest is required for this, you give the reader's an easy tool to figure out how to integrate your instructions with Jest's recognition of the unit tests. I know you posted a link, but just be sure.

I want to keep the guide laser focused on its topic -- it's meant to be a quick primer, after all, specifically for UI testing in Vue.
But I'll make sure to link the documentation and clarify it's done in the "usual way".

Firebrandt commented:
Okay. I wrote my first draft with your feedback, moved the primer to a better directory, removed the old one, and linked it properly as required. 

Hopefully I've integrated all your feedback.

The things I'm worried about now that the content I cover seems solid:
- Are there any glaring issues with the prose?
- Could it be a little too long for a quick primer, or does this seem about good to you?
- Are there any technical issues you see?

Fight-Fighter reviewed (commented):
First off, good that you added a link in the Tech_Stacks.md file.

**Preamble:**
I was going to say "add that you should do the command "npm install" but I think that's too granular. Otherwise, good section

**What's UI Testing?**
Neat, concise description.

**Why Test UI?**
You added in use cases like I asked, thank you.

**How do we test stuff?**
Thank you for specifying the different versions, it's important. Every other section is good too, there's just one major thing I'd like to request. Can you add code examples? I'm looking at the documentation for Vue tests and it has examples for how Vue testing functions are structured. Look here: https://testing-library.com/docs/vue-testing-library/examples
It would help to just include some simple ones to round the primer out a bit, explaining the importance of importing the components or the parameters in each function. Other than that, good stuff, Ill keep reading it over in case I think of something else that can be improved upon

Firebrandt commented:
Ok. I'll figure out the Tech_Stacks.md (probably by removing the Whitespace), and add a code example. 

I think the test we made for the accordion would work -- I'll add a few extra comments. (I don't want to copy directly from the docs).

Firebrandt commented:
OH OH OH! REMEMBER TO LINK TO THIS FROM A HIGHER LEVEL FILE.

Firebrandt commented:
Resolved the issue -- basically just had to tell git "yeah include the extra changes others have made".

Firebrandt commented:
I also added an example test (that we actually used!) with fairly verbose comments explaining what's happening.
